streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -416,6 +419,10 @@ private void shutdownTasksAndState() {
         if (cleanRun && firstException.get() == null) {
             firstException.set(commitOffsets());
         }
+        // Close all task producers
+        if (exactlyOnceEnabled) {
```

Comment:
The check here is mandatory -- the other check has only the purpose to guard against bugs.